Policies and tokens

emk opened the pull request:
Thank you for suggesting that I submit a PR! This PR contains a number of related changes, but each patch makes one, discrete change, and each patch has an explanation of why that change was made in the commit message. So it should be possible to step through the individual patches when reviewing.

This PR includes implementations of two-thirds of the APIs proposed in #5, including:
- Renewing arbitrary auth tokens (using `renew_token`), not just the client's own token (using `renew`). It's important to keep the existing API separate and not reimplement it terms of `renew/$TOKEN`, because some tokens have access to `renew-self` but not `renew/$TOKEN`.
- Listing the policies supported by vault. This normally requires `root` permission, but can be enabled using `sudo` policies if you're clever.

**Not yet implemented** in this PR:
- Creating new tokens. This API will require a lot of optional parameters, which we'll probably want to pass using a `Builder`-style struct. I'll work on this in the coming days.

Also included in this PR:
- Some minor updates and notes to help support Vault 0.6.1, and to make it easier for people to test against it. One of these changes **breaks API compatibility for the `Auth` struct**, so you may need to bump semver when you make a release.
- A bug fix for several existing APIs that I noticed while working on the new features. I've added doc tests for all the APIs that had this bug.

Please feel free to suggest changes and improvements to this patch! And thank you for taking the time to review it.

emk committed:
Upgrade test setup to use the latest version of vault

Unless there's a specific reason to test against 0.4.1, it might be good
to test against the latest build, which offers the latest APIs.

emk committed:
README.md: Set VAULT_ADDR to http when creating token

VAULT_ADDR now defaults to https://127.0.0.1:8200, which will fail
unless the user has set up SSL certs.

emk committed:
Get a policy list from /sys/policy

emk committed:
API BREAKING: Mark `Auth::metadata` as optional

Under vault 0.6.1, certain requests may return an `Auth` object with a
null `metadata` field.  This will probably require bumping the library version to
indicate a breaking API change for some users.

emk committed:
Factor out `increment` serialization

We'll need this for new APIs that we're about to add, so let's implement
it in just one place.

emk committed:
Implement renew_token for non-self tokens

This is useful for deployment tools that manage tokens for other
services.

emk committed:
Bug fix: Do not prepend `self.host` for `self.get` or `self.post`

While adding new features, I discovered that some of the existing APIs
for manipulating tokens and leases contained a bug where `self.get` and
related functions were prefixing `self.host` to URLs, but some of the
callers of `self.get` still thought they needed to prepend `self.host`
themselves.

So I fixed this issue everywhere that I could find it, and I added new
doc tests for as many of the affected APIs as possible.

The tests for two APIs are marked as `no_run`:

- `revoke` should not be run, because doing so would revoke our test
  token.
- `renew_lease` is currently disabled because the `get_secret` API
  currently offers no way to get a list ID for an existing secret.

emk committed:
Travis CI fix: Make sure our test token is renewable

This is needed to test the two token-renewal APIs.

emk commented:
Oops, I also need to make sure that Travis CI can test token renewal. I've added a new patch which should handle this case.

ChrisMacNaughton commented on src/client/mod.rs:
````diff
@@ -320,26 +336,75 @@ impl<'a, T> VaultClient<'a, T>
     /// Revoke `VaultClient`'s token. This token can no longer be used.
+    /// Corresponds to [`/auth/token/revoke-self`][token].
+    ///
+    /// ```no_run
````
Instead of making it no_run, what if we generate a new token as a part of this test?

emk replied:
I will update this test to do exactly that one I add a token-generation API, which is next on my list. :-)

ChrisMacNaughton replied:
Deal :smile:

ChrisMacNaughton commented on src/client/mod.rs:
````diff
@@ -320,26 +336,75 @@ impl<'a, T> VaultClient<'a, T>
+    /// Renew a specific lease that your token controls.  Corresponds to
+    /// [`/v1/sys/renew`][renew].
+    ///
+    /// ```no_run
````
As above, could we just create a new token to use for this example?

emk replied:
Ditto. Will be added when we have token creation support.

ChrisMacNaughton commented on src/client/mod.rs:
```diff
@@ -320,26 +336,75 @@ impl<'a, T> VaultClient<'a, T>
+    /// let token = "test12345";
+    /// let mut client = Client::new(host, token).unwrap();
+    ///
+    /// // TODO: Right now, we offer no way to get lease information for a
```
Is this something that is provided by Vault but not (correctly) implemented by us?

emk replied:
Yes. I've filed a separate issue in #7, which will also allow us to make this doc test runnable.

ChrisMacNaughton commented:
So, it looks like this passed on the 0.6 release, but failed on the older releases. Ideally, I'd like to support older releases but I'm open to breaking that requirement and tying the crate to Vault versions.

ChrisMacNaughton commented:
Maybe we can use some things with features, so that it can be compiled with less features by specifying specific flags somewhat like rustc does, ie: `features=vault-6.0` ?

emk commented:
I'm working on a fix for vault 0.4.1 as we speak. I don't think we need features for individual versions, at least not yet.

emk committed:
API BREAKING: Handle missing `accessor` in `Auth` (Vault < 0.6.1)

This field is not provided by earlier versions of Vault.  We can simply
make it optional and it should work fine.

As far as I can tell, this was actually broken before, but I don't think
that any of our pre-existing test cases actually hit it.  So by adding
new tests for various APIs that we fixed, we revealed underlying
problems with an older version of vault.

ChrisMacNaughton commented:
yeah, ideally we can support everything as a least common denominator, but I wouldn't be surprised of breaking changes in Vault that would require us to only enable features for versions >= x

emk commented:
OK, it looks like this Vault 0.4.1 issue was actually hiding already, and it was only revealed when I added more test cases and Travis CI ran them against Vault. I'm very pleased that you added support for testing against all supported vault versions automatically. :-)

For now, all the tested APIs work fine regardless of version. Vault's HTTP API has actually been quite stable for over a year, so we can probably get away with having one API and just testing against all configurations.

The only question will arise if we add new 0.6.1-specific features that can't work against older vault servers. At the very least, we would need to disable those tests when running against an older vault.

ChrisMacNaughton commented:
That idea of new features introduced in new versions is where I think features may come in very handy. Maybe make the default-features include everything, but we can modify the builds to handle disabling newer features as necessary.

emk commented:
Seems reasonable enough. I suppose Travis CI will let us know if we implement an API that doesn't work right under older Vault releases, and the feature can be added then.

emk commented:
OK, it looks like the entire build matrix is green on Travis CI!
